### PR TITLE
[MM-29240] Improve buffer handling of UploadEmojiImage

### PR DESCRIPTION
**Pull request by hanzei**
#### Summary

As an OSF project, I looked into the buffer usage in `UploadEmojiImage`.

The biggest improvements come from the removal of `io.Copy` to a new buffer. Instead, the same `io.ReadSeeker` is used in the whole method. Only if the image needs to be rescaled, a new buffer get allocated.

I also added a benchmark for confidence. As you can see, the improvements aren't huge, but they are there. Especially, for large gifs and images, memory consumption goes down by 30%.

In order to write the benchmark, I removed the dependency of `uploadEmojiImage` onto `multipart.FileHeader`. I thought about also replacing `*multipart.Form` in `CreateEmoji` with `io.ReadSeeker`, but that might be too much for this PR. Thoughts?

```
goos: linux
goarch: amd64
pkg: github.com/mattermost/mattermost/server/v8/channels/app
cpu: Intel(R) Core(TM) i7-9850H CPU @ 2.60GHz
                                              │   master.txt    │               MM-29240.txt           │
                                              │    sec/op        │    sec/op     vs base                │
UploadEmojiImage/gif/small-12                   54.49µ ±  16%   47.44µ ±  8%  -12.94% (p=0.009 n=10)
UploadEmojiImage/gif/max_size-12                61.11µ ±  13%   55.10µ ± 12%   -9.83% (p=0.005 n=10)
UploadEmojiImage/gif/too_wide-12                3.945m ±  15%   3.221m ±  6%  -18.35% (p=0.000 n=10)
UploadEmojiImage/gif/too_tall-12                4.411m ±  14%   3.211m ± 16%  -27.22% (p=0.000 n=10)
UploadEmojiImage/gif/too_tall_and_too_wide-12   26.87m ±   8%   21.61m ±  4%  -19.57% (p=0.000 n=10)
UploadEmojiImage/png/small-12                   67.37µ ± 120%   47.90µ ± 13%  -28.90% (p=0.000 n=10)
UploadEmojiImage/png/max_size-12                65.94µ ±  41%   57.87µ ±  9%  -12.23% (p=0.003 n=10)
UploadEmojiImage/png/too_wide-12                2.473m ±  14%   2.058m ± 15%  -16.80% (p=0.002 n=10)
UploadEmojiImage/png/too_tall-12                2.465m ±  11%   2.057m ± 11%  -16.58% (p=0.004 n=10)
UploadEmojiImage/png/too_tall_and_too_wide-12   13.80m ±  20%   11.81m ±  4%  -14.43% (p=0.000 n=10)
geomean                                         947.3µ          777.7µ        -17.90%

                                              │   master.txt   │              MM-29240.txt          │
                                              │     B/op         │     B/op      vs base               │
UploadEmojiImage/gif/small-12                   12.32Ki ± 0%   11.35Ki ± 0%  -7.86% (p=0.000 n=10)
UploadEmojiImage/gif/max_size-12                12.45Ki ± 0%   11.35Ki ± 0%  -8.79% (p=0.000 n=10)
UploadEmojiImage/gif/too_wide-12                1.061Mi ± 0%   1.027Mi ± 0%  -3.20% (p=0.000 n=10)
UploadEmojiImage/gif/too_tall-12                1.064Mi ± 0%   1.031Mi ± 0%  -3.18% (p=0.000 n=10)
UploadEmojiImage/gif/too_tall_and_too_wide-12   6.151Mi ± 0%   6.112Mi ± 0%  -0.64% (p=0.000 n=10)
UploadEmojiImage/png/small-12                   6.258Ki ± 0%   6.086Ki ± 0%  -2.75% (p=0.000 n=10)
UploadEmojiImage/png/max_size-12                6.555Ki ± 0%   6.086Ki ± 0%  -7.15% (p=0.000 n=10)
UploadEmojiImage/png/too_wide-12                1.650Mi ± 0%   1.644Mi ± 0%  -0.37% (p=0.000 n=10)
UploadEmojiImage/png/too_tall-12                1.648Mi ± 0%   1.640Mi ± 0%  -0.47% (p=0.000 n=10)
UploadEmojiImage/png/too_tall_and_too_wide-12   5.844Mi ± 0%   5.832Mi ± 0%  -0.20% (p=0.000 n=10)
geomean                                         245.7Ki        237.0Ki       -3.51%

                                              │  master.txt  │              MM-29240.txt           │
                                              │ allocs/op     │ allocs/op   vs base                │
UploadEmojiImage/gif/small-12                   281.0 ± 0%   278.0 ± 0%   -1.07% (p=0.000 n=10)
UploadEmojiImage/gif/max_size-12                281.0 ± 0%   278.0 ± 0%   -1.07% (p=0.000 n=10)
UploadEmojiImage/gif/too_wide-12                897.0 ± 0%   625.0 ± 0%  -30.32% (p=0.000 n=10)
UploadEmojiImage/gif/too_tall-12                897.0 ± 0%   625.0 ± 0%  -30.32% (p=0.000 n=10)
UploadEmojiImage/gif/too_tall_and_too_wide-12   898.0 ± 0%   625.0 ± 0%  -30.40% (p=0.000 n=10)
UploadEmojiImage/png/small-12                   23.00 ± 0%   20.00 ± 0%  -13.04% (p=0.000 n=10)
UploadEmojiImage/png/max_size-12                23.00 ± 0%   20.00 ± 0%  -13.04% (p=0.000 n=10)
UploadEmojiImage/png/too_wide-12                144.0 ± 0%   135.0 ± 0%   -6.25% (p=0.000 n=10)
UploadEmojiImage/png/too_tall-12                144.0 ± 0%   135.0 ± 0%   -6.25% (p=0.000 n=10)
UploadEmojiImage/png/too_tall_and_too_wide-12   144.0 ± 0%   135.0 ± 0%   -6.25% (p=0.000 n=10)
geomean                                         197.5        168.6       -14.61%
```
#### Ticket Link
https://mattermost.atlassian.net/browse/MM-29240

#### Release Note
```release-note
Improve performance of emoji upload
```

### CODE REVIEW

File: server/channels/app/emoji.go [hunk 1]
```diff
@@ -127,41 +117,42 @@ func (a *App) UploadEmojiImage(c request.CTX, id string, imageData *multipart.Fi
 		}, "", http.StatusBadRequest)
 	}
 
+	_, err = file.Seek(0, io.SeekStart)
+	if err != nil {
+		return model.NewAppError("uploadEmojiImage", "api.emoji.upload.seek.app_error", nil, "", http.StatusInternalServerError).Wrap(err)
+	}
+
+	var resized io.Reader = file
 	if config.Width > MaxEmojiWidth || config.Height > MaxEmojiHeight {
```
[the remark this answers is not in the document]
Comment (hanzei) [reply server/channels/app/emoji.go]: That's a great suggestion. Implemented via 658077dd0ae84a4a8da610eecfc007498bc5909d.

File: server/channels/app/emoji.go [hunk 2]
```diff
@@ -68,11 +68,18 @@ func (a *App) CreateEmoji(c request.CTX, sessionUserId string, emoji *model.Emoj
 		return nil, model.NewAppError("Context", "api.context.invalid_body_param.app_error", map[string]any{"Name": "createEmoji"}, "", http.StatusBadRequest)
 	}
 
-	if appErr := a.UploadEmojiImage(c, emoji.Id, imageData[0]); appErr != nil {
+	filename := imageData[0].Filename
+	file, err := imageData[0].Open()
+	if err != nil {
+		return nil, model.NewAppError("uploadEmojiImage", "api.emoji.upload.open.app_error", nil, "", http.StatusBadRequest).Wrap(err)
```
Comment (agarciamontoro) [inline server/channels/app/emoji.go:74]:
```suggestion
		return nil, model.NewAppError("createEmoji", "api.emoji.upload.open.app_error", nil, "", http.StatusBadRequest).Wrap(err)
```

**Review by agarciamontoro — APPROVED** [review]
It looks great! Just a nit: can we refactor the benchmark to avoid the repetition of the for loop in every test? We could simply use a slice of testcases with the different `utils.CreateTestXYZ` and dimensions, and then iterate over those and apply the same loop body to them.

File: server/channels/app/emoji.go (hunk 2 above)
Comment (hanzei) [reply server/channels/app/emoji.go:74]: Good catch :+1:

**Comment by hanzei** [comment]
> Just a nit: can we refactor the benchmark to avoid the repetition of the for loop in every test? We could simply use a slice of testcases with the different `utils.CreateTestXYZ` and dimensions, and then iterate over those and apply the same loop body to them.

While that is feasible, I'm hesitant as it makes it harder to understand the side effects of the benchmarks. One important thing for benchmarks is that they actually test the performance of the code you want and are not tainted by side effects. (see also https://dave.cheney.net/2013/06/30/how-to-write-benchmarks-in-go)

In this case: Correctness > verbosity

**Comment by hanzei** [comment]
@lindy65 Would you please do some smoke testing around custom emoji uploads with different file sizes? No behaviour changes are expected with this PR.

**Review by lindy65 — APPROVED** [review]
Thanks @hanzei - smoke-tested custom emoji and looks good to me 👍

**Pull request merged**